### PR TITLE
Fix UASocketProtocol receive_buffer

### LINKED ISSUE #40: I got an error when I got a lot of nodes  asyncua.ua.uaerrors._base.UaError: Wrong sequence 518 -> 518 (server bug or replay attack)
I got an error when I got a lot of nodes.
This error is not necessarily happening,It may take a few more runs to happen.
I have a device and thousands of nodes in my kepserver.

**stack trace**
```111Node(StringNodeId(ns=2;s=Collection))
222Node(StringNodeId(ns=2;s=Collection.ANDON1))
333Node(StringNodeId(ns=2;s=Collection.ANDON1.alarm))
Exception raised while parsing message from server
Traceback (most recent call last):
  File "C:\Python3\lib\site-packages\asyncua\client\ua_client.py", line 68, in _process_received_data
    msg = self._connection.receive_from_header_and_body(header, buf)
  File "C:\Python3\lib\site-packages\asyncua\common\connection.py", line 282, in receive_from_header_and_body
    return self._receive(chunk)
  File "C:\Python3\lib\site-packages\asyncua\common\connection.py", line 299, in _receive
    self._check_incoming_chunk(msg)
  File "C:\Python3\lib\site-packages\asyncua\common\connection.py", line 262, in _check_incoming_chunk
    .format(self._peer_sequence_number, num)
asyncua.ua.uaerrors._base.UaError: Wrong sequence 518 -> 518 (server bug or replay attack)
close_session was called but connection is closed
close_secure_channel was called but connection is closed
disconnect_socket was called but connection is closed
Traceback (most recent call last):
  File "C:/work/async/client-minimal.py", line 31, in <module>
    loop.run_until_complete(main())
  File "C:\Python3\lib\asyncio\base_events.py", line 484, in run_until_complete
    return future.result()
  File "C:/work/async/client-minimal.py", line 23, in main
    for d in await c.get_children():
  File "C:\Python3\lib\asyncio\coroutines.py", line 110, in __next__
    return self.gen.send(None)
  File "C:\Python3\lib\site-packages\asyncua\common\node.py", line 403, in get_referenced_nodes
    references = await self.get_references(refs, direction, nodeclassmask, includesubtypes)
  File "C:\Python3\lib\asyncio\coroutines.py", line 110, in __next__
    return self.gen.send(None)
  File "C:\Python3\lib\site-packages\asyncua\common\node.py", line 382, in get_references
    results = await self.server.browse(params)
  File "C:\Python3\lib\asyncio\coroutines.py", line 110, in __next__
    return self.gen.send(None)
  File "C:\Python3\lib\site-packages\asyncua\client\ua_client.py", line 300, in browse
    data = await self.protocol.send_request(request)
  File "C:\Python3\lib\asyncio\coroutines.py", line 110, in __next__
    return self.gen.send(None)
  File "C:\Python3\lib\site-packages\asyncua\client\ua_client.py", line 122, in send_request
    await asyncio.wait_for(future, timeout if timeout else None)
  File "C:\Python3\lib\asyncio\tasks.py", line 362, in wait_for
    raise futures.TimeoutError()
concurrent.futures._base.TimeoutError

```

### PR DESCRIPTION
**Pull request by cbergmiller**
Fixes #40
When `UASocketProtocol ._process_received_data()` had data left over after processing a message it set the `receive_buffer` to the originally received data. Instead `receive_buffer` must be set to what is left over after processing the message.
The whole buffering concept could use some cleanup and test coverage but if this fixes the issue we should merge and refactor in a new branch.

### CODE REVIEW

File: asyncua/client/ua_client.py [hunk 1]
```diff
@@ -77,8 +77,11 @@ def _process_received_data(self, data: bytes):
                     return
                 msg = self._connection.receive_from_header_and_body(header, buf)
                 self._process_received_message(msg)
-                if not buf:
+                if not len(buf):
+                    # Buffer is empty
```
Comment (oroulet) [inline asyncua/client/ua_client.py]: my understanding is that __bool__ is more pythonic. but feel free to arg against

File: asyncua/common/utils.py [hunk 2]
```diff
@@ -38,34 +38,30 @@ def __init__(self, data, start_pos=0, size=-1):
...
-            self._data[self._cur_pos:self._cur_pos + self._size])
+        return f"Buffer(size:{self._size}, data:{self._data[self._cur_pos:self._cur_pos + self._size]})"
     __repr__ = __str__
 
     def __len__(self):
         return self._size
 
-    def __bool__(self):
-        return self._size > 0
+    def __bytes__(self):
```
Comment (oroulet) [inline asyncua/common/utils.py]: that one is nice

File: examples/client-minimal.py [hunk 3]
```diff
@@ -4,12 +4,12 @@
...
 from asyncua import Client, Node, ua
 
-logging.basicConfig(level=logging.INFO)
+logging.basicConfig(level=logging.DEBUG)
 _logger = logging.getLogger('asyncua')
 
 
 async def main():
-    url = 'opc.tcp://localhost:4840/freeopcua/server/'
+    url = 'opc.tcp://192.168.2.64:4840'
```
Comment (oroulet) [inline examples/client-minimal.py]: do not harcode a special IP in examples. if yo do not want localhost use 127.0.0.1. Also the uri path is also common to give, although it is absolutely not used anywhere in code

**Comment by oroulet** [comment]
I do not know where the bug was, so I am not sure what it fixes, butI trust you

File: examples/client-minimal.py (hunk 3 above)
Comment (cbergmiller) [reply examples/client-minimal.py]: Did not mean to commit this - thanks for the catch!

**Comment by Ilyazyk** [comment]
Great job! It works so far for me.

**Comment by oroulet** [comment]
@cbergmiller you need to rebase and look at this __bool__ stuff

**Comment by cbergmiller** [comment]
I messed up that rebase somehow. I will open a new pull request.

**Pull request closed without merging**